google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java

```diff
@@ -81,17 +381,159 @@ public boolean equals(Object o) {
+    switch (type) {
+      case TYPE_UNSPECIFIED:
+        return com.google.spanner.admin.instance.v1.InstanceConfig.Type.TYPE_UNSPECIFIED;
+      case GOOGLE_MANAGED:
+        return com.google.spanner.admin.instance.v1.InstanceConfig.Type.GOOGLE_MANAGED;
+      case USER_MANAGED:
+        return InstanceConfig.Type.USER_MANAGED;
+      default:
+        throw new IllegalArgumentException("Unknown config type:" + type);
+    }
+  }
+
+  static InstanceConfig.State toProtoState(State state) {
+    switch (state) {
+      case STATE_UNSPECIFIED:
+        return com.google.spanner.admin.instance.v1.InstanceConfig.State.STATE_UNSPECIFIED;
+      case CREATING:
+        return com.google.spanner.admin.instance.v1.InstanceConfig.State.CREATING;
+      case READY:
+        return com.google.spanner.admin.instance.v1.InstanceConfig.State.READY;
+      default:
+        throw new IllegalArgumentException("Unknown state:" + state);
+    }
+  }
+
+  static com.google.cloud.spanner.InstanceConfig fromProto(
+      com.google.spanner.admin.instance.v1.InstanceConfig proto, InstanceAdminClient client) {
+    com.google.cloud.spanner.InstanceConfig.Builder builder =
+        (com.google.cloud.spanner.InstanceConfig.Builder)
+            new com.google.cloud.spanner.InstanceConfig.Builder(
+                    client, InstanceConfigId.of(proto.getName()))
+                .setReconciling(proto.getReconciling())
+                .setReplicas(
+                    proto.getReplicasList().stream()
+                        .map(ReplicaInfo::fromProto)
+                        .collect(Collectors.toList()))
+                .setDisplayName(proto.getDisplayName())
+                .putAllLabels(proto.getLabelsMap())
+                .setEtag(proto.getEtag())
+                .setLeaderOptions(proto.getLeaderOptionsList())
+                .setOptionalReplicas(
+                    proto.getOptionalReplicasList().stream()
+                        .map(ReplicaInfo::fromProto)
+                        .collect(Collectors.toList()))
+                .setState(fromProtoState(proto.getState()))
+                .setConfigType(fromProtoConfigType(proto.getConfigType()));
+
+    if (!proto.getBaseConfig().isEmpty()) {
+      builder.setBaseConfig(
+          new com.google.cloud.spanner.InstanceConfigInfo.BuilderImpl(
+                  InstanceConfigId.of(proto.getBaseConfig()))
+              .build());
+    }
+
+    return builder.build();
+  }
+
+  static State fromProtoState(com.google.spanner.admin.instance.v1.InstanceConfig.State state) {
+    switch (state) {
+      case STATE_UNSPECIFIED:
+        return State.STATE_UNSPECIFIED;
+      case CREATING:
+        return State.CREATING;
+      case READY:
+        return State.READY;
+      default:
+        throw new IllegalArgumentException("Unknown state:" + state);
+    }
+  }
+
+  static Type fromProtoConfigType(com.google.spanner.admin.instance.v1.InstanceConfig.Type type) {
```

Comment:
Could this be private? If it is more open just for testing, could you add the `@VisibleForTesting` annotation?

Reply:
Making it private makes sense. It is not directly used in unit tests, so I think we need not add `@VisibleForTesting` annotation.